x/ccv/provider/keeper/relay.go

```diff
@@ -147,47 +149,73 @@ func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data c
+func (k Keeper) HandleSlashPacket(ctx sdk.Context, chainID string, data ccv.SlashPacketData) error {
 	// map VSC ID to infraction height for the given chain ID
 	var infractionHeight uint64
-	if downtimeData.ValsetUpdateId == 0 {
+	if data.ValsetUpdateId == 0 {
 		infractionHeight = k.GetInitChainHeight(ctx, chainID)
 	} else {
-		infractionHeight = k.GetValsetUpdateBlockHeight(ctx, downtimeData.ValsetUpdateId)
+		infractionHeight = k.GetValsetUpdateBlockHeight(ctx, data.ValsetUpdateId)
 	}
 
+	// return if there isn't any initial chain height for the consumer chain
 	if infractionHeight == 0 {
-		return fmt.Errorf("cannot find validator update id %d for chain %s", downtimeData.ValsetUpdateId, chainID)
+		return fmt.Errorf("cannot find validator update id %d for chain %s", data.ValsetUpdateId, chainID)
 	}
 
-	// get the validator consensus address
-	consAddr := sdk.ConsAddress(downtimeData.Validator.Address)
-
-	// get the validator data
+	// get the validator
+	consAddr := sdk.ConsAddress(data.Validator.Address)
 	validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, consAddr)
-	if !found {
-		return fmt.Errorf("cannot find validator with address %s", consAddr.String())
-	}
 
 	// make sure the validator is not yet unbonded;
 	// stakingKeeper.Slash() panics otherwise
-	if validator.IsUnbonded() {
+	if !found || validator.IsUnbonded() {
 		return fmt.Errorf("should not be slashing unbonded validator: %s", validator.GetOperator())
 	}
 
-	// slash and jail the validator
-	k.stakingKeeper.Slash(ctx, consAddr, int64(infractionHeight), downtimeData.Validator.Power, sdk.NewDec(1).QuoInt64(downtimeData.SlashFraction))
-	if !validator.IsJailed() {
-		k.stakingKeeper.Jail(ctx, consAddr)
+	// spare jailed and/or tombstoned validator preventing to slash it again
+	if validator.IsJailed() || k.slashingKeeper.IsTombstoned(ctx, consAddr) {
+		return fmt.Errorf("should not be slashing jailed and/or tombstoned validator: %s", validator.GetOperator())
 	}
 
-	k.slashingKeeper.JailUntil(ctx, consAddr, ctx.BlockHeader().Time.Add(time.Duration(downtimeData.JailTime)))
+	// slash and jail validator according to their infraction type
+	// and using the provider chain parameters
+	var (
+		jailTime      time.Time
+		slashFraction sdk.Dec
+	)
+
+	switch data.Infraction {
+	// set the downtime slash fraction and duration
+	// then append the validator address to the slash ack for its chain id
+	case stakingtypes.Downtime:
+		slashFraction = k.slashingKeeper.SlashFractionDowntime(ctx)
+		jailTime = ctx.BlockTime().Add(k.slashingKeeper.DowntimeJailDuration(ctx))
+		k.AppendSlashAck(ctx, chainID, consAddr.String())
+	// set double-signing slash fraction and infinite jail duration
+	// then tombstone the validator
+	case stakingtypes.DoubleSign:
+		slashFraction = k.slashingKeeper.SlashFractionDoubleSign(ctx)
+		jailTime = evidencetypes.DoubleSignJailEndTime
+		k.slashingKeeper.Tombstone(ctx, consAddr)
+	default:
```

Comment:
I'm probably misunderstanding the protocol, but do you need to also send the ack in case of DoubleSign slash?

Reply:
`SlashAck`s prevent that a validator gets slashed multiple times on the provider chain for having downtime. This scenario can happen in CCV as pointed out in this [issue](https://github.com/cosmos/ibc/issues/685). You can also check the `SendSlashRequest`'s logic in the [ICS 26 spec],(https://github.com/cosmos/ibc/blob/marius/ccv/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-pcf-rcvslash1) where the consumer chains use `outstandingDowntime` to record and track the expected slash acknowledgments.

@mpoke [onRecvSlashPacket](https://github.com/cosmos/ibc/blob/marius/ccv/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-pcf-rcvslash1) needs to be updated in that regard.

Reply:
> onRecvSlashPacket needs to be updated in that regard.
@sainoe Do you mean calling `AppendSlashAck` only for downtime slashing?

Reply:
> You should still be sending acks even if you don't perform logic on them so sending chain can clean up their state

@AdityaSripal These SlashAcks here are not IBC Packet Acknowlegdments, but rather some info sent by the provider to the consumer (in the VSC packet) notifying it that the Slash requests for downtime were processed.

Reply:
> @sainoe Do you mean calling AppendSlashAck only for downtime slashing?`

@mpoke Yes!

Reply:
> Ahh ok, so double sign packets still get an acknowledgement?

Yes. 

> I assume the default acknowledgment is used?

Cannot use them because of the order with VSC packets.